lib/elixir/src/elixir_exp.erl

```diff
@@ -391,6 +403,20 @@ expand(Other, E) ->
 
 %% Helpers
 
+expand_boolean_check(Expr, TrueClause, FalseClause, Meta, Env) ->
+  {EExpr, EnvExpr} = expand(Expr, Env),
+  Clauses =
+    case elixir_utils:returns_boolean(EExpr) of
+      true ->
+        [TrueClause, FalseClause];
+      false ->
+        Other = {other, Meta, ?MODULE},
+        OtherExpr = {{'.', Meta, [erlang, error]}, Meta, [{badarg, Other}]},
```

Comment:
@lexmag that sounds good to me too. If so, this is good to merge. I will approve it. :)